[WIP] Add TypeScript 2.1 typings

sublimator opened the pull request:
These headers are exported from a branch where I converted the lib/**/*.js to TypeScript. The tests all passed, when I modified them slightly to make sure they referenced package.json `main` pointing to the tsconfig.compilerOptions.outDir.

sublimator commented:
re: #536 

I extended the client options from Url, because of the funny parsing done, but actually the `port` is more generally a `number` so need to fix that. 

Also TODO: remove the `_\w+` private functions

sublimator commented on typings/lib/client.d.ts:
```diff
@@ -0,0 +1,255 @@
+    on(event: 'error', cb: OnErrorCallback): this;
+    on(event: string, cb: Function): any;
+    once(event: 'message', cb: OnMessageCallback): this;
+    once(event: 'packetsend' | 'packetreceive', cb: OnPacketCallback): this;
```
![image](https://cloud.githubusercontent.com/assets/525211/23141413/4ae5903e-f7e9-11e6-9e7e-e671dde3924b.png)

![image](https://cloud.githubusercontent.com/assets/525211/23141432/5c855ab8-f7e9-11e6-8832-3c33bdd18752.png)

sublimator commented on typings/lib/client.d.ts:
```diff
@@ -0,0 +1,255 @@
+     * _reconnect - implement reconnection
+     * @api privateish
+     */
+    private _reconnect();
```
need to kill these off

sublimator commented on typings/lib/client-options.d.ts:
```diff
@@ -0,0 +1,113 @@
+/**
+ * MQTT CLIENT
+ */
+export interface ClientOptions extends SecureClientOptions, Url {
```
Should just inline the `Url` options ... playing around

sublimator commented:
https://github.com/Microsoft/TypeScript/issues/1867#issuecomment-209018980

Hrmmm

sublimator commented:
Removing the private members was my initial impulse also, but one of the m$
devs made a point that it helps when extending the class to not
accidentally over-ride private members. The tsc compiler will balk on any
attempts to use the members, even on extended classes, where `protected`
must be used.

Having said that, you could argue it's less a maintenance burden to only
expose the public API.

Thoughts?

I'm easy either way really. Out at the moment so can't make any changes.

One other thing to consider is how much of the `url.parse` options do we
actually want in the ClientOptions. The 'slashes' etc ... it's kinda messy.

RangerMauve commented:
With MQTT in the wild, I haven't ever needed to extend it directly, favoring composition over inheritance when adding new capabilities. I would vote for not exposing the private variables for simplicity. (for autocomplete, for example)

sublimator commented:
So it seems like the consensus is to remove the private variables. There was some discussion about having an option to omit them in the automatically generated .d.ts files (which is what these are) but seems like MS dug their heels in on this one. I'll just remove them manually. 

@mcollina As far as tests go, we'd need to have a typescript compiler added to the dev-dependencies, which is not really a biggie. What sort of testing did you have in mind? Just simple tests (connect, subscribe, publish, receive etc) or more extensive ones to somehow notice when the declarations ones get out of sync?

mcollina commented:
At this point, just making sure if those are valid it's good.

mcollina commented:
I mean linting them.

sublimator commented:
Yeah, that seems like a good simple step 👍 
Ok, I will remove the private variables, install `tslint` and add it to the CI build :) 

Probably tomorrow

sublimator commented on package.json:
```diff
@@ -84,6 +86,9 @@
     "through2": "^2.0.3",
+    "tslint": "^4.5.1",
+    "tslint-config-standard": "^4.0.0",
+    "typescript": "^2.2.1",
```
typescript is a peer dependency of tslint

sublimator commented on types/lib/client-options.d.ts:
```diff
@@ -0,0 +1,124 @@
+import { QoS } from './types'
+
+export interface IClientOptions extends ISecureClientOptions {
+  // The options are extended by url.parse() fields in connect() function
```
@mcollina 

What do you want exposed here? And once we've decided, should we only extend the IClientOptions with those, rather than 
just grabbing the kitchen sink?

```
> u.parse('mqtt://www.broker.com/mqtt')
Url {
  protocol: 'mqtt:',
  slashes: true,
  auth: null,
  host: 'www.broker.com',
  port: null,
  hostname: 'www.broker.com',
  hash: null,
  search: null,
  query: null,
  pathname: '/mqtt',
  path: '/mqtt',
  href: 'mqtt://www.broker.com/mqtt' }
```

I'll switch to TypeScript branch and `find usages` and report back

sublimator replied:
@mcollina I had a crack at cleaning up the url.parse/xtend section. Take a look

sublimator commented on types/lib/client.d.ts:
```diff
@@ -0,0 +1,190 @@
+ * (see Connection#connect)
+ */
+export declare class MqttClient extends events.EventEmitter {
+  // public connackTimer: NodeJS.Timer
```
these weren't prefixed with a `_` so I marked them public (default in TypeScript if not specifide) (the first tslint configuration, before I used the `standard` variant wanted explicit so I just used the --fix options) 

In any case, I guess we don't really want them exposed?

sublimator commented on lib/connect/index.js:
```diff
@@ -49,26 +81,16 @@ function connect (brokerUrl, opts) {
-    opts.clientId = opts.query.clientId
-  }
+  // Someone out in the wild might be using {auth: 'user:pass'} out in the wild
+  // Keep old behaviour? Tests actually pass without this.
```
Personally I'd just remove this ...

sublimator commented on lib/connect/index.js:
```diff
@@ -49,26 +81,16 @@ function connect (brokerUrl, opts) {
-    opts = xtend(parsed, opts)
-
-    if (opts.protocol === null) {
+    // Options object always override brokerUrl specified options
```
I did it this way due to a test:

```
    it('should throw an error when called with no protocol specified - with options', function () {
      (function () {
        mqtt.connect('tcp://foo.bar.com', { protocol: null })
      }).should.throw('Missing protocol')
    })
```

sublimator commented on lib/connect/index.js:
```diff
@@ -21,19 +21,51 @@ protocols.wss = require('./ws')
+      // protocol
+      // clientId
+    {
+      hostname: parsed.hostname,
```
It's interesting that the `servers` option uses {host, port} keys:

```
      opts.host = opts.servers[client._reconnectCount].host
      opts.port = opts.servers[client._reconnectCount].port
      opts.hostname = opts.host
```

Perhaps at some point it would be nice just to use one or the other

sublimator replied:
I mean use only one of host or hostname consistently

sublimator replied:
Hrmmm, ok, so I looked up the url.parse() function and

![image](https://cloud.githubusercontent.com/assets/525211/23782526/aad1636e-0587-11e7-9c81-9831b16c2bfa.png)

host is supposedly the hosname + ':' + port

sublimator commented:
@mcollina 

Ok, I removed private variables, and deprecated the `hostname` option in lieu of `host` which is used in the `servers: [{host: 'localhost', port: 1337},...]` option, for consistency. I also removed most of the jibberish url parse options (hash, query, etc), reducing it to just `port` (as a number) `host, path, protocol` and `clientId` and `username` / `password`.

sublimator commented:
The tests on Travis pass, but because of Js code changes related to hostname, the coverage has gone `down` by a fraction so it's marked as failed.

sublimator commented on lib/connect/index.js:
```diff
@@ -21,19 +21,57 @@ protocols.wss = require('./ws')
+      // protocol
+      // clientId
+    {
+      // hostname: parsed.hostname,
```
You could argue that to be consistent with deprecating, should leave the `hostname` until it's completely removed. Assuming you guys are down with that direction.

mcollina commented:
Can you please do not touch/remove/alter any existing functionality in this PR? I'm not sure what you want to achieve, and that is a semver-major change, while this is not.

I'll comment on the typescript questions later on.

sublimator commented:
My intention was to clean up the messy extending of the IClientOptions with
*ALL* the url.parse variables. It's quite glaring once you see it all
mapped out in TypeScript declarations, so I got a bit carried away making
alterations, while removing the url jibberish like `hash`/`query` etc. We
use hostname/host variously, while in server we use `host`.

I added a deprecation warning for `hostname`, for later removal but if
removing the rest warrants a sem-ver change then so be it. Is that an
issue? I can separate the changes into multiple pull requests if you'd like.

sublimator commented:
^ while in the *servers* config array we use `host`.

mcollina commented:
@sublimator are you ok in closing this? Or it's a mistake?

sublimator commented:
I will open another one later on :) Out at the moment
